source/includes/fundamentals/code-snippets/IndexPage.java

```diff
@@ -0,0 +1,147 @@
+        System.out.println(String.format("Index created: %s", resultCreateIndex));
+        // end multikey index
+
+        // begin covered multikey query
+        Bson filter = and(eq("genres", "Animation"), eq("rated", "G"));
+        Bson sort = Sorts.ascending("title");
+        Bson projection = fields(include("title", "rated"), excludeId());
+        FindIterable<Document> cursor = collection.find(filter).sort(sort).projection(projection);
+        // end covered multikey query
+        cursor.forEach(doc -> System.out.println(doc));
+    }
+
+    private void textIndex() {
+        System.out.println("text index");
+        // begin text index
+        try {
+            String resultCreateIndex = collection.createIndex(Indexes.text("plot"));
+            System.out.println(String.format("Index created: %s", resultCreateIndex));
+        } catch (MongoCommandException e) {
+            if (e.getErrorCodeName().equals("IndexOptionsConflict"))
+                System.out.println("there's an existing text index with different options");
+        }
+        // end text index
+
+        // begin text query
+        Bson filter = text("java coffee shop");
+        Bson projection = fields(include("fullplot"), excludeId());
+        FindIterable<Document> cursor = collection.find(filter).projection(projection);
+        // end text query
+        cursor.forEach(doc -> System.out.println(doc));
+    }
+
+    private void geoSpatialIndex() {
+        System.out.println("geospatial index");
+        collection = database.getCollection("theaters");
+        // begin geospatial index
+        try {
+            String resultCreateIndex = collection.createIndex(Indexes.geo2dsphere("location.geo"));
+            System.out.println(String.format("Index created: %s", resultCreateIndex));
+        } catch (MongoCommandException e) {
+            if (e.getErrorCodeName().equals("IndexOptionsConflict"))
+                System.out.println("there's an existing geo index with different options");
+        }
+        // end geospatial index
+
+        // begin geospatial query
+        // MongoDB Headquarters in NY, NY.
+        Point refPoint = new Point(new Position(-73.98456, 40.7612));
+        Bson filter = near("location.geo", refPoint, 1000.0, 0.0);
+        FindIterable<Document> cursor = collection.find(filter);
+        // end geospatial query
+        cursor.forEach(doc -> System.out.println(doc));
+    }
+
+    private void uniqueIndex() {
+        System.out.println("unique index");
+        collection = database.getCollection("theaters");
+        // begin unique index
+        try {
+            IndexOptions indexOptions = new IndexOptions().unique(true);
+            String resultCreateIndex = collection.createIndex(Indexes.descending("theaderId"), indexOptions);
+            System.out.println(String.format("Index created: %s", resultCreateIndex));
+        } catch (DuplicateKeyException e) {
```

Comment:
Even if I create the index twice in a row, I don't get a DuplicateKeyException (or any exception for that matter).  Is MongoDB just treating the second attempt as a no-op?

Reply:
The DuplicateKeyException would be thrown if you tried to create an index with unique constraints on a field that [isn't unique](https://docs.mongodb.com/manual/core/index-unique/#restrictions).

```java
    private void uniqueIndex() {
        System.out.println("unique index");
        collection = database.getCollection("uniqueViolator");

        collection.insertMany(Arrays.asList(new Document("a", 1), new Document("a", 1)));
        // begin unique index
        // this will fail if any duplicate keys (values) exist
        try {
            IndexOptions indexOptions = new IndexOptions().unique(true);
            String resultCreateIndex = collection.createIndex(Indexes.descending("a"), indexOptions);
            System.out.println(String.format("Index created: %s", resultCreateIndex));
        } catch (DuplicateKeyException e) {
            System.out.println("duplicate key, couldn't create index");
        }
        // end unique index
    }
```